lib/cdo/redis.rb

```diff
@@ -0,0 +1,121 @@
+require 'redis'
+require 'singleton'
+require 'json'
+require 'cdo/buffer'
+
+# Wrapper client for Redis in memory store
+class RedisClient
+  include Singleton
+
+  # :i18n
+  # Stream for tracking which translated strings are used on which URLs.
+  # @example
+  #   RedisClient.instance.put_record(
+  #     :i18n,
+  #     {
+  #       url: 'http://studio.code.org/courses',
+  #       string_key: 'home.heading_elementary_markdown'
+  #     }
+  #   )
+
+  STREAMS = {
+    i18n: "i18n-string-tracking-events"
+  }.freeze
+
+  cattr_accessor :client
+  cattr_accessor :log
+
+  # 'The max length of a list is 2^32 - 1 elements (4294967295, more than 4 billion of elements per list).'
+  # Ref: https://redis.io/topics/data-types#:~:text=The%20max%20length%20of%20a%20list%20is%20232%20-%201%20elements%20(4294967295,%20more%20than%204%20billion%20of%20elements%20per%20list).
+  MAX_LIST_LENGTH = 2**32 - 1
+
+  class Buffer < Cdo::Buffer
```

Comment:
If the `Buffer` class here and in `FirehoseClient` are mostly the same, can we extract it as a shared class between 2 clients? https://github.com/code-dot-org/code-dot-org/blob/cc9c363c627eb534acd417befa266416fb7a26ea/lib/cdo/firehose.rb#L64

Reply:
Or you can give it a different name, `RedisBuffer` for example.

Reply:
Cleaning up the FirehoseClient buffer used to be tracked in the Jira item to swap the two usages, but it's been removed: https://codedotorg.atlassian.net/browse/FND-1539?atlOrigin=eyJpIjoiY2YyMmE3MDk2YWE5NDlmMjk5OTQ4MjYyMzVkMWQ1MGQiLCJwIjoiaiJ9

In the meantime I can rename the buffer class